website_sale_product_publish_date/models/product_template.py

```diff
@@ -0,0 +1,36 @@
+# Copyright 2024 Quartile (https://www.quartile.co)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+from odoo.osv import expression
+
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    expected_publish_date = fields.Datetime()
+    expected_unpublish_date = fields.Datetime()
+
+    def _get_publish_date_domain(self):
+        """Generate domain for publish and unpublish date filtering."""
+        now = fields.Datetime.now()
+        return [
+            "|",
+            "&",
+            ("expected_publish_date", "<=", now),
+            "|",
+            ("expected_unpublish_date", ">=", now),
+            ("expected_unpublish_date", "=", False),
+            "&",
+            ("expected_publish_date", "=", False),
+            "|",
+            ("expected_unpublish_date", ">=", now),
+            ("expected_unpublish_date", "=", False),
+        ]
+
+    @api.model
+    def _search_build_domain(self, domain_list, search, fields, extra=None):
+        base_domain = super()._search_build_domain(domain_list, search, fields, extra)
+        publish_domain = self._get_publish_date_domain()
+        combined_domain = expression.AND([base_domain, publish_domain])
```

Comment:
We should do this only on the product model.

It may also be safer to check `website_published` is included in `domain_list`.